rdflib/namespace/__init__.py

```diff
@@ -239,6 +239,16 @@ def __dir__(cls) -> Iterable[str]:
         values = {cls[str(x)] for x in cls.__annotations__}
         return values
 
+    def jsonld_context(self, pfx: str) -> str:
+        """This function creates a JSON-LD 'context' JSON object from this
+        DefinedNamespace namespace and it's members"""
+        terms = {pfx: str(self._NS)}
+        for key, term in self.__annotations__.items():
+            if term == URIRef:
```

Comment:
Not entirely sure there is a good use-case for this, but may be better to do this in case someone subclasses `URIRef`, though I think other things will also fall apart if it is a subclass so maybe it does not really add value:

```suggestion
            if issubclass(term, URIRef):
```

Reply:
Shouldn't we use `isinstance` instead of `issubclass`?

Reply:
Yes!

Reply:
Because we are testing an object, not a class.

Reply:
this is an annotation though, and `__annotations__` for a class is a dict where the keys are the names of the attributes, and the values are the types of the attributes:

`tests/test_annotations.py`:

```python
class URIRef:
    pass


class URIRefB(URIRef):
    pass


class MyClass:
    a: URIRef
    b: URIRefB


def test_types() -> None:
    assert issubclass(URIRefB, URIRef)
    assert not isinstance(URIRefB, URIRef)


def test_annotations() -> None:
    assert MyClass.__annotations__.keys() == {"a", "b"}

    for key, term in MyClass.__annotations__.items():
        assert isinstance(key, str)
        assert issubclass(term, URIRef)
        assert not isinstance(term, URIRef)

    ann = MyClass.__annotations__
    assert ann["a"] == URIRef
    assert issubclass(ann["a"], URIRef)
    assert not issubclass(ann["a"], URIRefB)
    assert not isinstance(ann["a"], URIRef)

    assert ann["b"] == URIRefB
    assert issubclass(ann["b"], URIRef)
    assert issubclass(ann["b"], URIRefB)
    assert not isinstance(ann["b"], URIRef)
    assert not isinstance(ann["b"], URIRefB)
```

```console
$ pipx run --spec pytest -- pytest tests/test_annotations.py -v
============================================================================ test session starts ============================================================================
platform linux -- Python 3.10.2, pytest-7.0.1, pluggy-1.0.0 -- /home/iwana/.local/pipx/.cache/bd575df7c04465a/bin/python
cachedir: .pytest_cache
rootdir: /home/iwana/sw/d/gitlab.com/aucampia/scratchpad/tech/py3, configfile: pyproject.toml
collected 2 items                                                                                                                                                           

tests/test_annotations.py::test_types PASSED                                                                                                                          [ 50%]
tests/test_annotations.py::test_annotations PASSED                                                                                                                    [100%]

============================================================================= 2 passed in 0.02s =============================================================================
```

Reply:
In this case also, changing this line to `isinstance` will result in test failure:

```console
$ git diff
diff --git a/rdflib/namespace/__init__.py b/rdflib/namespace/__init__.py
index ab0170f2..5460cd4c 100644
--- a/rdflib/namespace/__init__.py
+++ b/rdflib/namespace/__init__.py
@@ -244,7 +244,7 @@ class DefinedNamespaceMeta(type):
         DefinedNamespace namespace and it's members"""
         terms = {pfx: str(self._NS)}
         for key, term in self.__annotations__.items():
-            if term == URIRef:
+            if isinstance(term, URIRef):
                 terms[key] = f'{pfx}:{key}'
 
         return json.dumps({'@context': terms})
$ .venv/bin/python3 -m pytest test/test_definednamespace.py --log-level DEBUG -rA --tb=short
============================================================================ test session starts ============================================================================
platform linux -- Python 3.9.10, pytest-7.0.0, pluggy-1.0.0
rootdir: /home/iwana/sw/d/github.com/iafork/rdflib.cleanish, configfile: tox.ini
plugins: subtests-0.6.0, cov-3.0.0
collected 5 items                                                                                                                                                           

test/test_definednamespace.py ....F                                                                                                                                   [100%]

================================================================================= FAILURES ==================================================================================
___________________________________________________________________ test_definednamespace_jsonld_context ____________________________________________________________________
test/test_definednamespace.py:174: in test_definednamespace_jsonld_context
    assert actual == expected
E   assert '{"@context":...skos/core#"}}' == '{"@context":...Collection"}}'
E     Skipping 50 identical leading characters in diff, use -v to show
E     - kos/core#", "altLabel": "skos:altLabel", "broadMatch": "skos:broadMatch", "broader": "skos:broader", "broaderTransitive": "skos:broaderTransitive", "changeNote": "skos:changeNote", "closeMatch": "skos:closeMatch", "definition": "skos:definition", "editorialNote": "skos:editorialNote", "exactMatch": "skos:exactMatch", "example": "skos:example", "hasTopConcept": "skos:hasTopConcept", "hiddenLabel": "skos:hiddenLabel", "historyNote": "skos:historyNote", "inScheme": "skos:inScheme", "mappingRelation": "skos:m...
E     
E     ...Full output truncated (2 lines hidden), use '-vv' to show
================================================================================== PASSES ===================================================================================
_____________________________________________________________________ test_definednamespace_creator_qb ______________________________________________________________________
--------------------------------------------------------------------------- Captured stdout call ----------------------------------------------------------------------------


Using /home/iwana/sw/d/github.com/iafork/rdflib.cleanish/rdflib/tools/defined_namespace_creator.py...
Testing /home/iwana/sw/d/github.com/iafork/rdflib.cleanish/test/defined_namespaces/qb.ttl...
____________________________________________________________________ test_definednamespace_creator_fake _____________________________________________________________________
--------------------------------------------------------------------------- Captured stdout call ----------------------------------------------------------------------------


Using /home/iwana/sw/d/github.com/iafork/rdflib.cleanish/rdflib/tools/defined_namespace_creator.py...
Testing /home/iwana/sw/d/github.com/iafork/rdflib.cleanish/test/defined_namespaces/fake.xxx...(expected to fail)
___________________________________________________________________ test_definednamespace_creator_bad_ns ____________________________________________________________________
--------------------------------------------------------------------------- Captured stdout call ----------------------------------------------------------------------------


Using /home/iwana/sw/d/github.com/iafork/rdflib.cleanish/rdflib/tools/defined_namespace_creator.py...
Testing /home/iwana/sw/d/github.com/iafork/rdflib.cleanish/test/defined_namespaces/fake.xxx...(expected to fail - bad NS given)
========================================================================== short test summary info ==========================================================================
PASSED test/test_definednamespace.py::test_definednamespace_creator_qb
PASSED test/test_definednamespace.py::test_definednamespace_creator_fake
PASSED test/test_definednamespace.py::test_definednamespace_creator_bad_ns
PASSED test/test_definednamespace.py::test_definednamespace_dir
FAILED test/test_definednamespace.py::test_definednamespace_jsonld_context - assert '{"@context":...skos/core#"}}' == '{"@context":...Collection"}}'
======================================================================== 1 failed, 4 passed in 0.78s ========================================================================
```

Reply:
However, with issubclass, it still passes:

```console
$ git diff
diff --git a/rdflib/namespace/__init__.py b/rdflib/namespace/__init__.py
index ab0170f2..6b49610f 100644
--- a/rdflib/namespace/__init__.py
+++ b/rdflib/namespace/__init__.py
@@ -244,7 +244,7 @@ class DefinedNamespaceMeta(type):
         DefinedNamespace namespace and it's members"""
         terms = {pfx: str(self._NS)}
         for key, term in self.__annotations__.items():
-            if term == URIRef:
+            if issubclass(term, URIRef):
                 terms[key] = f'{pfx}:{key}'
 
         return json.dumps({'@context': terms})
20220212T194153 iwana@iwana-pc00.coop.no:~/sw/d/github.com/iafork/rdflib.cleanish
$ .venv/bin/python3 -m pytest test/test_definednamespace.py --log-level DEBUG -rA --tb=short
============================================================================ test session starts ============================================================================
platform linux -- Python 3.9.10, pytest-7.0.0, pluggy-1.0.0
rootdir: /home/iwana/sw/d/github.com/iafork/rdflib.cleanish, configfile: tox.ini
plugins: subtests-0.6.0, cov-3.0.0
collected 5 items                                                                                                                                                           

test/test_definednamespace.py .....                                                                                                                                   [100%]

================================================================================== PASSES ===================================================================================
_____________________________________________________________________ test_definednamespace_creator_qb ______________________________________________________________________
--------------------------------------------------------------------------- Captured stdout call ----------------------------------------------------------------------------


Using /home/iwana/sw/d/github.com/iafork/rdflib.cleanish/rdflib/tools/defined_namespace_creator.py...
Testing /home/iwana/sw/d/github.com/iafork/rdflib.cleanish/test/defined_namespaces/qb.ttl...
____________________________________________________________________ test_definednamespace_creator_fake _____________________________________________________________________
--------------------------------------------------------------------------- Captured stdout call ----------------------------------------------------------------------------


Using /home/iwana/sw/d/github.com/iafork/rdflib.cleanish/rdflib/tools/defined_namespace_creator.py...
Testing /home/iwana/sw/d/github.com/iafork/rdflib.cleanish/test/defined_namespaces/fake.xxx...(expected to fail)
___________________________________________________________________ test_definednamespace_creator_bad_ns ____________________________________________________________________
--------------------------------------------------------------------------- Captured stdout call ----------------------------------------------------------------------------


Using /home/iwana/sw/d/github.com/iafork/rdflib.cleanish/rdflib/tools/defined_namespace_creator.py...
Testing /home/iwana/sw/d/github.com/iafork/rdflib.cleanish/test/defined_namespaces/fake.xxx...(expected to fail - bad NS given)
========================================================================== short test summary info ==========================================================================
PASSED test/test_definednamespace.py::test_definednamespace_creator_qb
PASSED test/test_definednamespace.py::test_definednamespace_creator_fake
PASSED test/test_definednamespace.py::test_definednamespace_creator_bad_ns
PASSED test/test_definednamespace.py::test_definednamespace_dir
PASSED test/test_definednamespace.py::test_definednamespace_jsonld_context
============================================================================= 5 passed in 0.71s =============================================================================
```